### PR TITLE
Creates Torch precision tests, moves more of test_cuda.py to test_torch.py

### PR DESCRIPTION
**Pull request by mruberry**
- Creates DevicePrecision in test_torch.py for tests that compare device results to CPU results
- Moves many tests from test_cuda.py to generic test_torch.py classes
- Adds slowTest decorator to two slow (empirically) slow tests
- Adds @requiresMemory decorator
- Lets generic tests suites be created only on devices (not on the CPU)

The test updates fix a few minor bugs and expose an unexpected gap. test_digamma, for example, originally stated that it was checking behavior of float tensors, but this is incorrect and the test is failing (at least for me) with float tensors.

### CODE REVIEW

**Comment by mruberry** [comment]
ONNX failures are occuring on Master. XLA failure is because XLA needs to exclude test_binary_op_mem_overlap_xla and test_ternary_op_mem_overlap_xla. Not sure why those aren't excluded. I'll review the tests shortly.

**Review by izdeby — APPROVED** [review]

File: test/test_torch.py [hunk 1]
```diff
@@ -8775,8 +8326,8 @@ def validate_setting(x):
 
         # test invalid index fails
         reference = torch.empty(10, device=device)
-        # can't test cuda because it is a device assert
-        if not reference.is_cuda:
+        # only tests asserts on cpu
+        if 'cpu' in device:
```
Comment (gchanan) [inline test/test_torch.py]: `torch.tensor([0], device=device).device == 'cpu'` (you can create a wrapper function for this if you want.

**Review by gchanan — COMMENTED** [review]
umm...can you ghstack this?  There's a bunch of code moves and new functionality in the same PR and it's difficult to see what's going on.

File: test/test_torch.py (hunk 1 above)
Comment (mruberry) [reply test/test_torch.py]: Sure you don't want to just pass actual devices?

Yeah I'll cut this PR up. Git doesn't do a great job showing these diffs.

**Comment by mruberry** [comment]
Splitting this for readability.

**Pull request closed without merging**